Refactor EuTxManager Java to Kotlin

serim53 opened the pull request:
**Refactor EuTxManager java to kotlin**
* Related Issue : #224 
- Convert Java style code to Kotlin style (Kotlin block function)
- Change Handler to Coroutine Scope
- Versionize AudioTrack
    - Apply Builder Pattern
- Deprecate setMode() and convert into Kotlin format

yeonns reviewed (commented):
@serim53 Thank you for refactoring EuTxManager! It looks good :+1: 
One thing I wanna comment is that `txCore` is  included at constructor. `txCore` is only used at `EuTxManager` internally, and its initialization is always same, so I think it would be removed at constructor. How about declaraing as `private val`?

And, `EuTxManagerTest` has unused imports, so how about removing it as well? :smile:

serim53 commented:
> removed at

As you said, txcore is only used internally, so I also think it's better to move it to the internal code! I corrected this part and deleted the import, and I commited it again! Thank you for checking carefully! 🥰

serim53 commented:
> @serim53 Thank you for refactoring EuTxManager! It looks good 👍 One thing I wanna comment is that `txCore` is included at constructor. `txCore` is only used at `EuTxManager` internally, and its initialization is always same, so I think it would be removed at constructor. How about declaraing as `private val`?
> 
> And, `EuTxManagerTest` has unused imports, so how about removing it as well? 😄

As you said, txcore is only used internally, so I also think it's better to move it to the internal code! I corrected this part and deleted the import, and I commited it again! Thank you for checking carefully! 🥰

designe reviewed (commented):
Hi! @serim53 
Sorry for the late riview 😢 

I read all your `EuTxManager.kt` code.
- By replacing Handler with corountine, android os dependency has disappeared, and flexibility seems to have been strengthened.
- In addition, readability enhancement was noticeable through AudioTrack's Builder pattern.

Overall, the refactored code seems to be neatly written.
I leave some review-comments here :)

SeonJK commented:
Great contribution!

I read your codes and have some questions.

In deprecated functions; setMode() and setCode(), is it also call txCore.setMode() and txCore.setCode()?

serim53 commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,173 @@
+        )
+        val bufferSize = outStream.size * minBufferSizeBytes
+        audioTrack = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            AudioTrack.Builder()
+                .apply {
+                    setAudioFormat(AudioFormat.Builder().apply {
+                        setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
+                        setSampleRate(Constants.SAMPLERATE)
+                        setEncoding(AudioFormat.ENCODING_PCM_FLOAT)
+                    }.build())
+                    setAudioAttributes(AudioAttributes.Builder().apply {
+                        setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
+                    }.build())
+                    setBufferSizeInBytes(bufferSize)
+                    setTransferMode(AudioTrack.MODE_STATIC)
+                }.build()
```
> Great contribution!
> 
> I read your codes and have some questions.
> 
> In deprecated functions; setMode() and setCode(), does it also call txCore.setMode() and txCore.setCode()?

<img width="254" alt="스크린샷 2022-09-14 오전 12 35 00" src="https://user-images.githubusercontent.com/67946662/189944539-5b905aa7-152f-4402-bea8-5d0723eb4059.png">

Yes, you are right! txCore.setMode() and txCore.setCode() are called by variable!

serim53 replied:
> Actually, Previous code has some `deprecated` constructor usage issue. It is resolved by `AudioTrack.Builder()` 👍
> 
> Actually, `Builder` pasttern greatly enhances readability. all set function of `Builder` returns `AudioTrack` value.
> 
> So below is also working well without `apply`.
> 
> ```kotlin
> AudioTrack.Builder()
>     .setAudioFormat(
>         AudioFormat.Builder()
>             .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
>             .setSampleRate(Constants.SAMPLERATE)
>             .setEncoding(AudioFormat.ENCODING_PCM_FLOAT)
>             .build()
>     )
>     .setAudioAttributes(
>         AudioAttributes.Builder()
>             .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
>             .build()
>     )
>     .setBufferSizeInBytes(bufferSize)
>     .setTransferMode(AudioTrack.MODE_STATIC)
>     .build()
> ```

It's neater! I changed it! 🙌🏻

serim53 commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,173 @@
+            CoroutineScope(Dispatchers.IO).launch {
+                delay(if (duration == EuPIDuration.LENGTH_SHORT) 200 else 500.toLong())
+                stop()
+            }
```
> It would be nice to change `Handler` to `corountine`.
> 
> I wonder why 200 and 500 are specified as different types. left one is `Int` & right one is `Long` type. actually, delay's parameter type is `Long`. If the purpose is to specify the type, below is also fine :)

Thank you for your review! I didn't know the type was set like that, I changed it in this commit!
315c48c

zion830 reviewed (commented):
The first Kotlin Transformation! 👍
I left some reviews about nullable :D

serim53 commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,173 @@
+    )
+    fun process(count: Int) {
+        play(count, PlayerEngine.ANDROID_DEFAULT_ENGINE)
+    }
```
Thank you for searching! 🥰
I modified at 2e62a9f !

zion830 commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,185 @@
+    /*
+	 * @deprecated Replaced by {@link #setModeType(String)}, deprecated for naming & dynamic option.
+	 */
+    @Deprecated("", ReplaceWith(
+        "modeType = type",
+        "co.euphony.util.EuOption.ModeType"
+        )
+    )
+    fun setMode(type: ModeType?) {
+        modeType = type
+    }
+
+    /*
+	 * @deprecated Replaced by {@link #setCode(String)}, deprecated for naming & dynamic option.
+	 */
+    @Deprecated("", ReplaceWith(
+        "code = data",
+        "co.euphony.common.EuNativeConnector"
```
I think you should use `setModeType(type)` and `setCode(data)` instead of `modeType = type` and `code = data`
The first argument should be a code piece that you want to replace. [(A post about this)](https://dev.to/mreichelt/the-hidden-kotlin-gem-you-didn-t-think-you-ll-love-deprecations-with-replacewith-3blo)

serim53 replied:
Thank you for your reference link! I could understand the reference. I modified it and committed it! 👍🏻

designe commented:
@zion830 Thanks for your helpful review :)
@serim53 It's almost done ;) after modificaiton about `@Deprecated` annotation, It's done!

serim53 commented:
> @zion830 Thanks for your helpful review :) @serim53 It's almost done ;) after modificaiton about `@Deprecated` annotation, It's done!

I finally modified all reviews! Thanks for your help!☺️

yeonns reviewed (approved):
@serim53 @zion830 Thank you for update! Now, it looks perfect!

designe commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,185 @@
+    /*
+	 * @deprecated Replaced by {@link #setCode(String)}, deprecated for naming & dynamic option.
+	 */
+    @Deprecated("", ReplaceWith(
+        "setCode(data)",
+        "co.euphony.common.EuNativeConnector"
+
+        )
+    )
```
```suggestion
    @Deprecated("Use the `code` property.")
```

`ReplaceWith` does not support the properties of getter and setter roles yet, so here is using a message only.

designe replied:
Actually, `setCode` function is not exists for kotlin. only working for java.

designe commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,185 @@
+        code = data
+    }
+
+    /*
+	 * @deprecated Replaced by {@link #play()}, deprecated for naming issue
+	 */
+    @Deprecated(
+        "", ReplaceWith(
+            "play(1, PlayerEngine.ANDROID_DEFAULT_ENGINE)",
+            "co.euphony.tx.EuTxManager.PlayerEngine"
+        )
+    )
+    fun process() {
+        play(1, PlayerEngine.ANDROID_DEFAULT_ENGINE)
+    }
```
```suggestion
    /*
     * @deprecated Replaced by {@link #play()}, deprecated for naming issue
     */
    @Deprecated("Use play() function instead", ReplaceWith("play()"))
    fun process() {
        play()
    }
```

designe commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,185 @@
+        play(1, PlayerEngine.ANDROID_DEFAULT_ENGINE)
+    }
+
+    /*
+	 * @deprecated Replaced by {@link #play(int)}, deprecated for naming issue
+	 */
+    @Deprecated(
+        "", ReplaceWith(
+            "play(count, PlayerEngine.ANDROID_DEFAULT_ENGINE)",
+            "co.euphony.tx.EuTxManager.PlayerEngine"
+        )
+    )
+    fun process(count: Int) {
+        play(count, PlayerEngine.ANDROID_DEFAULT_ENGINE)
+    }
```
```suggestion
    /*
     * @deprecated Replaced by {@link #play(int)}, deprecated for naming issue
     */
    @Deprecated("Use play(count: int) function simply", ReplaceWith("play(count)"))
    fun process(count: Int) {
        play(count)
    }
```

designe commented on euphony/src/main/java/co/euphony/tx/EuTxManager.kt:
```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    /*
+	 * @deprecated Replaced by {@link #setModeType(String)}, deprecated for naming & dynamic option.
+	 */
+    @Deprecated("", ReplaceWith(
+        "setModeType(type)",
+        "co.euphony.util.EuOption.ModeType"
+        )
+    )
+    fun setMode(type: ModeType?) {
+        modeType = type
+    }
```
```suggestion
    /*
     * @deprecated Replaced by {@link #setModeType(String)}, deprecated for naming.
     */
    @Deprecated("Use the `modeType` property")
    fun setMode(type: ModeType?) {
        modeType = type
    }
```
`ReplaceWith` does not support the properties of getter and setter roles yet, so here is using a message only.

designe replied:
Actually, `setModeType` function is not exists for kotlin. only working for java.

designe reviewed (approved):
Since `ReplaceWith` does not yet support the properties of getter and setter roles, it seems that there is currently only a way to display it with a message.
For the rest, the same expression was expressed more concisely.

@serim53 Thanks for your great migration & refactoring!
